File: validators/python/rcip_validator.py
#!/usr/bin/env python3
"""
RCIP Format Validator for Python
Created by Alexey Kozlov
Version: 1.0.0
License: MIT
"""

import json
import re
import sys
from pathlib import Path
from typing import Dict, List, Optional, Tuple, Any
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
import jsonschema
from jsonschema import validate, ValidationError, Draft7Validator


class DietLabel(Enum):
    """Valid diet labels"""
    VEGETARIAN = "vegetarian"
    VEGAN = "vegan"
    GLUTEN_FREE = "gluten-free"
    DAIRY_FREE = "dairy-free"
    NUT_FREE = "nut-free"
    EGG_FREE = "egg-free"
    SOY_FREE = "soy-free"
    FISH_FREE = "fish-free"
    SHELLFISH_FREE = "shellfish-free"
    KOSHER = "kosher"
    HALAL = "halal"
    LOW_SODIUM = "low-sodium"
    LOW_CARB = "low-carb"
    KETO = "keto"
    PALEO = "paleo"


class Allergen(Enum):
    """Valid allergens"""
    MILK = "milk"
    EGGS = "eggs"
    FISH = "fish"
    SHELLFISH = "shellfish"
    TREE_NUTS = "tree-nuts"
    PEANUTS = "peanuts"
    WHEAT = "wheat"
    GLUTEN = "gluten"
    SOYBEANS = "soybeans"
    SESAME = "sesame"
    CELERY = "celery"
    MUSTARD = "mustard"
    MOLLUSCS = "molluscs"
    LUPINS = "lupins"
    SULPHITES = "sulphites"
    LACTOSE = "lactose"


class CookingAction(Enum):
    """Valid cooking actions"""
    ADD = "add"
    MIX = "mix"
    COMBINE = "combine"
    BLEND = "blend"
    CUT = "cut"
    SLICE = "slice"
    DICE = "dice"
    CHOP = "chop"
    MINCE = "mince"
    HEAT = "heat"
    BOIL = "boil"
    SIMMER = "simmer"
    STEAM = "steam"
    FRY = "fry"
    SAUTE = "saute"
    BAKE = "bake"
    ROAST = "roast"
    GRILL = "grill"
    COOL = "cool"
    CHILL = "chill"
    FREEZE = "freeze"
    KNEAD = "knead"
    FOLD = "fold"
    ROLL = "roll"
    SHAPE = "shape"
    FERMENT = "ferment"
    PROOF = "proof"
    REST = "rest"
    STRAIN = "strain"
    FILTER = "filter"
    SEPARATE = "separate"
    MEASURE = "measure"
    WEIGH = "weigh"
    WAIT = "wait"
    DISSOLVE = "dissolve"
    PREPARE = "prepare"
    SPREAD = "spread"
    GARNISH = "garnish"
    DIVIDE = "divide"


class Unit(Enum):
    """Valid measurement units"""
    # Mass
    MG = "mg"
    G = "g"
    KG = "kg"
    OZ = "oz"
    LB = "lb"
    # Volume
    ML = "ml"
    L = "l"
    TSP = "tsp"
    TBSP = "tbsp"
    CUP = "cup"
    FL_OZ = "fl-oz"
    PT = "pt"
    QT = "qt"
    GAL = "gal"
    # Count
    PCS = "pcs"
    DOZEN = "dozen"
    # Special
    PINCH = "pinch"
    DASH = "dash"
    HANDFUL = "handful"
    TO_TASTE = "to-taste"


@dataclass
class ValidationResult:
    """Validation result structure"""
    valid: bool
    errors: List[str] = field(default_factory=list)
    warnings: List[str] = field(default_factory=list)
    info: Dict[str, Any] = field(default_factory=dict)


@dataclass
class ValidationStats:
    """Validation statistics"""
    validated: int = 0
    passed: int = 0
    failed: int = 0


class RCIPValidator:
    """RCIP Format Validator"""

    RECIPE_ID_PATTERN = re.compile(r'^rcip-[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$')
    INGREDIENT_ID_PATTERN = re.compile(r'^ing-[0-9a-zA-Z]+$')
    STEP_ID_PATTERN = re.compile(r'^s-[0-9a-zA-Z]+$')
    VERSION_PATTERN = re.compile(r'^\d+\.\d+\.\d+$')
    COUNTRY_CODE_PATTERN = re.compile(r'^[A-Z]{2}$')

    def __init__(self, schema_version: str = "0.1"):
        """Initialize validator with schema version"""
        self.schema_version = schema_version
        self.schema = None
        self.validator = None
        self.stats = ValidationStats()

    def init(self, schema_path: Optional[Path] = None) -> None:
        """
        Initialize validator with schema

        Args:
            schema_path: Path to JSON schema file
        """
        if schema_path is None:
            schema_path = Path(__file__).parent.parent.parent / "schemas" / f"rcip-v{self.schema_version}.json"

        try:
            with open(schema_path, 'r', encoding='utf-8') as f:
                self.schema = json.load(f)

            self.validator = Draft7Validator(self.schema)
            print(f"✅ RCIP Validator initialized with schema v{self.schema_version}")
        except FileNotFoundError:
            raise FileNotFoundError(f"Schema file not found: {schema_path}")
        except json.JSONDecodeError as e:
            raise ValueError(f"Invalid JSON in schema file: {e}")

    def validate_recipe(self, recipe: Dict[str, Any]) -> ValidationResult:
        """
        Validate a recipe object

        Args:
            recipe: Recipe dictionary to validate

        Returns:
            ValidationResult object
        """
        if self.validator is None:
            raise RuntimeError("Validator not initialized. Call init() first.")

        result = ValidationResult(valid=True)

        # JSON Schema validation
        try:
            self.validator.validate(recipe)
        except ValidationError as e:
            result.valid = False
            result.errors.append(f"Schema validation failed: {e.message}")
            # Collect all errors
            for error in self.validator.iter_errors(recipe):
                error_path = " > ".join(str(p) for p in error.path)
                result.errors.append(f"{error_path}: {error.message}")

        # Custom validations
        self._validate_custom_rules(recipe, result)

        # Check warnings
        result.warnings = self._check_warnings(recipe)

        # Get recipe info
        result.info = self._get_recipe_info(recipe)

        # Update stats
        self.stats.validated += 1
        if result.valid:
            self.stats.passed += 1
        else:
            self.stats.failed += 1

        return result

    def validate_file(self, file_path: Path) -> ValidationResult:
        """
        Validate a recipe file

        Args:
            file_path: Path to recipe file

        Returns:
            ValidationResult object
        """
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                recipe = json.load(f)

            print(f"\n📄 Validating: {file_path.name}")
            result = self.validate_recipe(recipe)

            self._print_result(result, recipe.get('meta', {}).get('name', 'Unknown Recipe'))
            return result

        except FileNotFoundError:
            return ValidationResult(
                valid=False,
                errors=[f"File not found: {file_path}"]
            )
        except json.JSONDecodeError as e:
            return ValidationResult(
                valid=False,
                errors=[f"Invalid JSON: {e}"]
            )

    def validate_directory(self, directory: Path) -> List[Dict[str, Any]]:
        """
        Validate all recipe files in a directory

        Args:
            directory: Directory path

        Returns:
            List of validation results
        """
        recipe_files = list(directory.glob("*.rcip")) + list(directory.glob("*.json"))
        print(f"\n🔍 Found {len(recipe_files)} recipe files to validate\n")

        results = []
        for file_path in recipe_files:
            result = self.validate_file(file_path)
            results.append({
                "file": file_path.name,
                "result": result
            })

        self._print_summary()
        return results

    def _validate_custom_rules(self, recipe: Dict[str, Any], result: ValidationResult) -> None:
        """Apply custom validation rules beyond JSON schema"""

        # Validate recipe ID format
        if 'id' in recipe and not self.RECIPE_ID_PATTERN.match(recipe['id']):
            result.valid = False
            result.errors.append(f"Invalid recipe ID format: {recipe['id']}")

        # Validate ingredients
        for i, ingredient in enumerate(recipe.get('ingredients', [])):
            self._validate_ingredient(ingredient, i, result)

        # Validate steps
        for i, step in enumerate(recipe.get('steps', [])):
            self._validate_step(step, i, result)

        # Validate cross-references
        self._validate_references(recipe, result)

        # Validate version compatibility
        if 'rcip_version' in recipe and recipe['rcip_version'] != self.schema_version:
            result.warnings.append(
                f"Recipe version {recipe['rcip_version']} may not be fully compatible with validator version {self.schema_version}")

    def _validate_ingredient(self, ingredient: Dict[str, Any], index: int, result: ValidationResult) -> None:
        """Validate an ingredient"""

        # Check ID format
        if 'id' in ingredient and not self.INGREDIENT_ID_PATTERN.match(ingredient['id']):
            result.valid = False
            result.errors.append(f"Ingredient {index}: Invalid ID format: {ingredient['id']}")

        # Check allergens (must be present, can be empty)
        if 'allergens' not in ingredient:
            result.valid = False
            result.errors.append(f"Ingredient {index}: Missing required allergens field")
        elif not isinstance(ingredient['allergens'], list):
            result.valid = False
            result.errors.append(f"Ingredient {index}: allergens must be an array")
        else:
            # Validate allergen values
            valid_allergens = [a.value for a in Allergen]
            for allergen in ingredient['allergens']:
                if allergen not in valid_allergens:
                    result.errors.append(f"Ingredient {index}: Invalid allergen '{allergen}'")

        # Check machine amount
        if 'machine_amount' in ingredient:
            ma = ingredient['machine_amount']
            if 'value' not in ma or not isinstance(ma['value'], (int, float)) or ma['value'] < 0:
                result.errors.append(f"Ingredient {index}: machine_amount.value must be non-negative number")
            if 'unit' not in ma:
                result.errors.append(f"Ingredient {index}: machine_amount.unit is required")
            elif ma['unit'] not in [u.value for u in Unit]:
                result.warnings.append(f"Ingredient {index}: Non-standard unit '{ma['unit']}'")

    def _validate_step(self, step: Dict[str, Any], index: int, result: ValidationResult) -> None:
        """Validate a step"""

        # Check ID format
        if 'step_id' in step and not self.STEP_ID_PATTERN.match(step['step_id']):
            result.valid = False
            result.errors.append(f"Step {index}: Invalid ID format: {step['step_id']}")

        # Check action
        if 'action' in step:
            valid_actions = [a.value for a in CookingAction]
            if step['action'] not in valid_actions:
                result.errors.append(f"Step {index}: Invalid action '{step['action']}'")

        # Check hazards
        if 'hazards' in step and isinstance(step['hazards'], list):
            valid_hazards = ['hot-surface', 'sharp-tool', 'electrical', 'chemical', 'pressure',
                             'allergen-cross-contact']
            for hazard in step['hazards']:
                if hazard not in valid_hazards:
                    result.warnings.append(f"Step {index}: Non-standard hazard '{hazard}'")

    def _validate_references(self, recipe: Dict[str, Any], result: ValidationResult) -> None:
        """Validate cross-references between steps and ingredients"""

        ingredient_ids = {ing['id'] for ing in recipe.get('ingredients', []) if 'id' in ing}
        step_ids = {step['step_id'] for step in recipe.get('steps', []) if 'step_id' in step}

        # Check step targets
        for step in recipe.get('steps', []):
            if 'target' in step and isinstance(step['target'], list):
                for target in step['target']:
                    if isinstance(target, str):
                        if target.startswith('ing-') and target not in ingredient_ids:
                            result.errors.append(
                                f"Step {step.get('step_id', '?')}: Invalid ingredient reference '{target}'")
                        elif ':result' in target:
                            step_ref = target.split(':')[0]
                            if step_ref not in step_ids:
                                result.errors.append(
                                    f"Step {step.get('step_id', '?')}: Invalid step reference '{target}'")

        # Check device profile references
        device_ids = {dp['id'] for dp in recipe.get('device_profiles', []) if 'id' in dp}
        for step in recipe.get('steps', []):
            if 'device_profile_ref' in step and step['device_profile_ref'] not in device_ids:
                result.warnings.append(
                    f"Step {step.get('step_id', '?')}: Unknown device profile '{step['device_profile_ref']}'")

    def _check_warnings(self, recipe: Dict[str, Any]) -> List[str]:
        """Check for warnings (non-critical issues)"""
        warnings = []

        meta = recipe.get('meta', {})

        # Check for missing optional but recommended fields
        if 'description' not in meta:
            warnings.append("Missing recommended field: meta.description")

        if 'servings' not in meta:
            warnings.append("Missing recommended field: meta.servings")

        if 'difficulty' not in meta:
            warnings.append("Missing recommended field: meta.difficulty")

        # Check for missing nutritional data
        has_nutritional = any(
            'nutritional' in ing
            for ing in recipe.get('ingredients', [])
        )
        if not has_nutritional:
            warnings.append("No nutritional data provided for any ingredient")

        # Check for missing external IDs
        has_external_ids = any(
            ing.get('external_ids', {})
            for ing in recipe.get('ingredients', [])
        )
        if not has_external_ids:
            warnings.append("No external IDs (USDA, GTIN, etc.) provided")

        # Check for device profiles without sensors
        if recipe.get('device_profiles') and not recipe.get('sensors'):
            warnings.append("Device profiles defined but no sensors specified")

        # Check for very long cooking times
        total_time = meta.get('total_time_minutes', 0)
        if total_time > 1440:  # 24 hours
            warnings.append(f"Very long cooking time ({total_time} min / {total_time / 60:.1f} hours)")

        # Check for missing images
        if not recipe.get('images'):
            warnings.append("No images provided for recipe")

        return warnings

    def _get_recipe_info(self, recipe: Dict[str, Any]) -> Dict[str, Any]:
        """Get recipe information summary"""
        meta = recipe.get('meta', {})

        # Get all allergens
        allergens = set()
        for ingredient in recipe.get('ingredients', []):
            for allergen in ingredient.get('allergens', []):
                allergens.add(allergen)

        return {
            'name': meta.get('name', 'Unknown'),
            'version': recipe.get('rcip_version'),
            'recipe_version': meta.get('version'),
            'ingredient_count': len(recipe.get('ingredients', [])),
            'step_count': len(recipe.get('steps', [])),
            'has_device_profiles': bool(recipe.get('device_profiles')),
            'has_sensors': bool(recipe.get('sensors')),
            'allergens': sorted(list(allergens)),
            'diet_labels': meta.get('diet_labels', []),
            'difficulty': meta.get('difficulty', 'not specified'),
            'total_time': meta.get('total_time_minutes', 'not specified')
        }

    def _print_result(self, result: ValidationResult, recipe_name: str) -> None:
        """Print validation result"""
        print("\n" + "=" * 60)
        print(f"Recipe: {recipe_name}")
        print(f"Status: {'✅ VALID' if result.valid else '❌ INVALID'}")

        if result.info:
            print("\n📊 Recipe Info:")
            print(f"  - RCIP Version: {result.info['version']}")
            print(f"  - Ingredients: {result.info['ingredient_count']}")
            print(f"  - Steps: {result.info['step_count']}")
            print(f"  - Difficulty: {result.info['difficulty']}")
            print(f"  - Total Time: {result.info['total_time']} minutes")

            if result.info['allergens']:
                print(f"  - Allergens: {', '.join(result.info['allergens'])}")

            if result.info['diet_labels']:
                print(f"  - Diet Labels: {', '.join(result.info['diet_labels'])}")

        if result.errors:
            print(f"\n❌ Errors ({len(result.errors)}):")
            for i, error in enumerate(result.errors[:10], 1):
                print(f"  {i}. {error}")
            if len(result.errors) > 10:
                print(f"  ... and {len(result.errors) - 10} more errors")

        if result.warnings:
            print(f"\n⚠️  Warnings ({len(result.warnings)}):")
            for warning in result.warnings:
                print(f"  - {warning}")

        print("=" * 60)

    def _print_summary(self) -> None:
        """Print validation summary"""
        print("\n" + "=" * 60)
        print("📈 VALIDATION SUMMARY")
        print("=" * 60)
        print(f"Total Validated: {self.stats.validated}")

        if self.stats.validated > 0:
            pass_rate = (self.stats.passed / self.stats.validated) * 100
            fail_rate = (self.stats.failed / self.stats.validated) * 100
            print(f"✅ Passed: {self.stats.passed} ({pass_rate:.0f}%)")
            print(f"❌ Failed: {self.stats.failed} ({fail_rate:.0f}%)")

        print("=" * 60 + "\n")

    def reset_stats(self) -> None:
        """Reset validation statistics"""
        self.stats = ValidationStats()


def main():
    """CLI interface"""
    import argparse

    parser = argparse.ArgumentParser(
        description='RCIP Format Validator',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  python rcip_validator.py recipe.rcip
  python rcip_validator.py recipes/
  python rcip_validator.py --version 0.2 recipe.json
        """
    )

    parser.add_argument(
        'target',
        type=Path,
        help='Recipe file or directory to validate'
    )

    parser.add_argument(
        '--version',
        default='0.1',
        help='RCIP schema version (default: 0.1)'
    )

    parser.add_argument(
        '--schema',
        type=Path,
        help='Path to custom schema file'
    )

    args = parser.parse_args()

    # Initialize validator
    validator = RCIPValidator(schema_version=args.version)

    try:
        validator.init(schema_path=args.schema)

        if args.target.is_dir():
            validator.validate_directory(args.target)
        elif args.target.is_file():
            validator.validate_file(args.target)
        else:
            print(f"Error: {args.target} is not a valid file or directory")
            sys.exit(1)

    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: validators/javascript/rcip-validator.js
/**
 * RCIP Format Validator for JavaScript/Node.js
 * Created by Alexey Kozlov
 * Version: 1.0.0
 * License: MIT
 */

const Ajv = require('ajv');
const addFormats = require('ajv-formats');
const fs = require('fs').promises;
const path = require('path');

/**
 * RCIP Validator Class
 * Validates recipes against the RCIP schema specification
 */
class RCIPValidator {
  constructor(schemaVersion = '0.1') {
    this.schemaVersion = schemaVersion;
    this.ajv = new Ajv({ allErrors: true, verbose: true });
    addFormats(this.ajv);
    this.schema = null;
    this.validate = null;
    this.stats = {
      validated: 0,
      passed: 0,
      failed: 0
    };
  }

  /**
   * Initialize validator with schema
   * @param {string} schemaPath - Path to JSON schema file
   */
  async init(schemaPath = null) {
    try {
      if (!schemaPath) {
        schemaPath = path.join(__dirname, '../../schemas', `rcip-v${this.schemaVersion}.json`);
      }

      const schemaContent = await fs.readFile(schemaPath, 'utf8');
      this.schema = JSON.parse(schemaContent);
      this.validate = this.ajv.compile(this.schema);

      console.log(`✅ RCIP Validator initialized with schema v${this.schemaVersion}`);
      return true;
    } catch (error) {
      console.error(`❌ Failed to initialize validator: ${error.message}`);
      throw error;
    }
  }

  /**
   * Validate a recipe object
   * @param {Object} recipe - Recipe object to validate
   * @returns {Object} Validation result
   */
  validateRecipe(recipe) {
    if (!this.validate) {
      throw new Error('Validator not initialized. Call init() first.');
    }

    const valid = this.validate(recipe);
    this.stats.validated++;

    if (valid) {
      this.stats.passed++;
    } else {
      this.stats.failed++;
    }

    return {
      valid,
      errors: valid ? null : this.validate.errors,
      warnings: this.checkWarnings(recipe),
      info: this.getRecipeInfo(recipe)
    };
  }

  /**
   * Validate a recipe file
   * @param {string} filePath - Path to recipe file
   * @returns {Object} Validation result
   */
  async validateFile(filePath) {
    try {
      const content = await fs.readFile(filePath, 'utf8');
      const recipe = JSON.parse(content);

      console.log(`\n📄 Validating: ${path.basename(filePath)}`);
      const result = this.validateRecipe(recipe);

      this.printResult(result, recipe.meta?.name || 'Unknown Recipe');
      return result;
    } catch (error) {
      console.error(`❌ Error reading file: ${error.message}`);
      return {
        valid: false,
        errors: [{ message: error.message }]
      };
    }
  }

  /**
   * Validate multiple recipe files
   * @param {string} directory - Directory containing recipe files
   * @returns {Object} Batch validation results
   */
  async validateDirectory(directory) {
    try {
      const files = await fs.readdir(directory);
      const recipeFiles = files.filter(f => f.endsWith('.rcip') || f.endsWith('.json'));

      console.log(`\n🔍 Found ${recipeFiles.length} recipe files to validate\n`);

      const results = [];
      for (const file of recipeFiles) {
        const filePath = path.join(directory, file);
        const result = await this.validateFile(filePath);
        results.push({ file, ...result });
      }

      this.printSummary();
      return results;
    } catch (error) {
      console.error(`❌ Error reading directory: ${error.message}`);
      throw error;
    }
  }

  /**
   * Check for warnings (non-critical issues)
   * @param {Object} recipe - Recipe object
   * @returns {Array} Warning messages
   */
  checkWarnings(recipe) {
    const warnings = [];

    // Check for missing optional but recommended fields
    if (!recipe.meta?.description) {
      warnings.push('Missing recommended field: meta.description');
    }

    if (!recipe.meta?.servings) {
      warnings.push('Missing recommended field: meta.servings');
    }

    if (!recipe.meta?.difficulty) {
      warnings.push('Missing recommended field: meta.difficulty');
    }

    // Check for missing nutritional data
    const hasNutritionalData = recipe.ingredients?.some(ing => ing.nutritional);
    if (!hasNutritionalData) {
      warnings.push('No nutritional data provided for any ingredient');
    }

    // Check for missing external IDs
    const hasExternalIds = recipe.ingredients?.some(ing => ing.external_ids && Object.keys(ing.external_ids).length > 0);
    if (!hasExternalIds) {
      warnings.push('No external IDs (USDA, GTIN, etc.) provided for ingredients');
    }

    // Check for device profiles without sensors
    if (recipe.device_profiles?.length > 0 && !recipe.sensors?.length) {
      warnings.push('Device profiles defined but no sensors specified');
    }

    // Check for very long cooking times
    if (recipe.meta?.total_time_minutes > 1440) { // 24 hours
      warnings.push('Very long cooking time (>24 hours) - ensure this is intentional');
    }

    // Check for missing images
    if (!recipe.images || recipe.images.length === 0) {
      warnings.push('No images provided for recipe');
    }

    return warnings;
  }

  /**
   * Get recipe information summary
   * @param {Object} recipe - Recipe object
   * @returns {Object} Recipe info
   */
  getRecipeInfo(recipe) {
    return {
      name: recipe.meta?.name,
      version: recipe.rcip_version,
      recipeVersion: recipe.meta?.version,
      ingredientCount: recipe.ingredients?.length || 0,
      stepCount: recipe.steps?.length || 0,
      hasDeviceProfiles: !!(recipe.device_profiles?.length > 0),
      hasSensors: !!(recipe.sensors?.length > 0),
      allergens: this.getAllAllergens(recipe),
      dietLabels: recipe.meta?.diet_labels || [],
      difficulty: recipe.meta?.difficulty || 'not specified',
      totalTime: recipe.meta?.total_time_minutes || 'not specified'
    };
  }

  /**
   * Get all allergens in recipe
   * @param {Object} recipe - Recipe object
   * @returns {Array} Unique allergens
   */
  getAllAllergens(recipe) {
    const allergens = new Set();

    recipe.ingredients?.forEach(ingredient => {
      ingredient.allergens?.forEach(allergen => {
        allergens.add(allergen);
      });
    });

    return Array.from(allergens);
  }

  /**
   * Validate specific components
   */

  validateIngredient(ingredient) {
    const errors = [];

    // Check required fields
    if (!ingredient.id) errors.push('Missing required field: id');
    if (!ingredient.name) errors.push('Missing required field: name');
    if (!ingredient.human_amount) errors.push('Missing required field: human_amount');
    if (!ingredient.machine_amount) errors.push('Missing required field: machine_amount');
    if (!Array.isArray(ingredient.allergens)) {
      errors.push('Missing required field: allergens (must be array, can be empty)');
    }

    // Validate ID format
    if (ingredient.id && !ingredient.id.match(/^ing-[0-9a-zA-Z]+$/)) {
      errors.push(`Invalid ingredient ID format: ${ingredient.id}`);
    }

    // Validate machine amount
    if (ingredient.machine_amount) {
      if (typeof ingredient.machine_amount.value !== 'number' || ingredient.machine_amount.value < 0) {
        errors.push('machine_amount.value must be a non-negative number');
      }
      if (!ingredient.machine_amount.unit) {
        errors.push('machine_amount.unit is required');
      }
    }

    // Validate allergens
    const validAllergens = [
      'milk', 'eggs', 'fish', 'shellfish', 'tree-nuts', 'peanuts',
      'wheat', 'gluten', 'soybeans', 'sesame', 'celery', 'mustard',
      'molluscs', 'lupins', 'sulphites', 'lactose'
    ];

    ingredient.allergens?.forEach(allergen => {
      if (!validAllergens.includes(allergen)) {
        errors.push(`Invalid allergen: ${allergen}`);
      }
    });

    return { valid: errors.length === 0, errors };
  }

  validateStep(step) {
    const errors = [];

    // Check required fields
    if (!step.step_id) errors.push('Missing required field: step_id');
    if (!step.human_text) errors.push('Missing required field: human_text');
    if (!step.action) errors.push('Missing required field: action');

    // Validate ID format
    if (step.step_id && !step.step_id.match(/^s-[0-9a-zA-Z]+$/)) {
      errors.push(`Invalid step ID format: ${step.step_id}`);
    }

    // Validate action
    const validActions = [
      'add', 'mix', 'combine', 'blend', 'cut', 'slice', 'dice', 'chop', 'mince',
      'heat', 'boil', 'simmer', 'steam', 'fry', 'saute', 'bake', 'roast', 'grill',
      'cool', 'chill', 'freeze', 'knead', 'fold', 'roll', 'shape', 'ferment',
      'proof', 'rest', 'strain', 'filter', 'separate', 'measure', 'weigh', 'wait',
      'dissolve', 'prepare', 'spread', 'garnish', 'divide'
    ];

    if (step.action && !validActions.includes(step.action)) {
      errors.push(`Invalid action: ${step.action}`);
    }

    // Validate hazards if present
    const validHazards = [
      'hot-surface', 'sharp-tool', 'electrical', 'chemical', 'pressure', 'allergen-cross-contact'
    ];

    step.hazards?.forEach(hazard => {
      if (!validHazards.includes(hazard)) {
        errors.push(`Invalid hazard: ${hazard}`);
      }
    });

    return { valid: errors.length === 0, errors };
  }

  /**
   * Print validation result
   * @param {Object} result - Validation result
   * @param {string} recipeName - Recipe name
   */
  printResult(result, recipeName) {
    console.log(`\n${'='.repeat(60)}`);
    console.log(`Recipe: ${recipeName}`);
    console.log(`Status: ${result.valid ? '✅ VALID' : '❌ INVALID'}`);

    if (result.info) {
      console.log(`\n📊 Recipe Info:`);
      console.log(`  - RCIP Version: ${result.info.version}`);
      console.log(`  - Ingredients: ${result.info.ingredientCount}`);
      console.log(`  - Steps: ${result.info.stepCount}`);
      console.log(`  - Difficulty: ${result.info.difficulty}`);
      console.log(`  - Total Time: ${result.info.totalTime} minutes`);
      if (result.info.allergens.length > 0) {
        console.log(`  - Allergens: ${result.info.allergens.join(', ')}`);
      }
      if (result.info.dietLabels.length > 0) {
        console.log(`  - Diet Labels: ${result.info.dietLabels.join(', ')}`);
      }
    }

    if (result.errors && result.errors.length > 0) {
      console.log(`\n❌ Errors (${result.errors.length}):`);
      result.errors.slice(0, 10).forEach((error, i) => {
        console.log(`  ${i + 1}. ${error.schemaPath || error.instancePath}: ${error.message}`);
      });
      if (result.errors.length > 10) {
        console.log(`  ... and ${result.errors.length - 10} more errors`);
      }
    }

    if (result.warnings && result.warnings.length > 0) {
      console.log(`\n⚠️  Warnings (${result.warnings.length}):`);
      result.warnings.forEach(warning => {
        console.log(`  - ${warning}`);
      });
    }

    console.log(`${'='.repeat(60)}`);
  }

  /**
   * Print validation summary
   */
  printSummary() {
    console.log(`\n${'='.repeat(60)}`);
    console.log('📈 VALIDATION SUMMARY');
    console.log(`${'='.repeat(60)}`);
    console.log(`Total Validated: ${this.stats.validated}`);
    console.log(`✅ Passed: ${this.stats.passed} (${Math.round(this.stats.passed/this.stats.validated*100)}%)`);
    console.log(`❌ Failed: ${this.stats.failed} (${Math.round(this.stats.failed/this.stats.validated*100)}%)`);
    console.log(`${'='.repeat(60)}\n`);
  }

  /**
   * Reset statistics
   */
  resetStats() {
    this.stats = {
      validated: 0,
      passed: 0,
      failed: 0
    };
  }
}

// Export for use in Node.js
module.exports = RCIPValidator;

// CLI Interface
if (require.main === module) {
  const validator = new RCIPValidator();

  async function main() {
    const args = process.argv.slice(2);

    if (args.length === 0) {
      console.log('Usage: node rcip-validator.js <file.rcip|directory>');
      process.exit(1);
    }

    try {
      await validator.init();

      const target = args[0];
      const stats = await fs.stat(target);

      if (stats.isDirectory()) {
        await validator.validateDirectory(target);
      } else {
        await validator.validateFile(target);
      }
    } catch (error) {
      console.error(`Error: ${error.message}`);
      process.exit(1);
    }
  }

  main();
}